File: src/main/java/com/leetcode/top/interview/easy/LongestCommonPrefix.java
package com.leetcode.top.interview.easy;

/**
 * 原想法是最菜的一种，现在采用分而治之+缓存
 *
 * @author wezhyn
 * @since 04.02.2020
 */
public class LongestCommonPrefix {




    public String longestCommonPrefix(String[] strs) {
        if (strs.length==1) {
            return strs[0];
        }
        String maxCom="";
        int maxLen=Integer.MAX_VALUE;
        for (int sep=1; sep < strs.length; sep+=sep) {
            for (int i=0; i + sep < strs.length; i+=2*sep) {
                String cur=strs[i];
                String next=strs[i + sep];
                final String prefix=commonPrefix(cur, next, maxLen);
                if (prefix.length() < maxLen) {
                    maxCom=prefix;
                    maxLen=prefix.length();
                }
            }
        }
        return maxCom;
    }

    public String commonPrefix(String one, String ano, int maxLen) {
        int len=Math.min(
                Math.min(ano.length() - 1, one.length() - 1),
                maxLen
        );
        int i;
        for (i=0; i <= len; i++) {
            if (one.charAt(i)!=ano.charAt(i)) {
                return one.substring(0, i);
            }
        }
        return ano.substring(0, i);
    }
}

File: leetcode-kotlin/src/main/kotlin/com/wezhyn/learn/string/LongestCommonPrefix.kt
package com.wezhyn.learn.string

/**
 * @see com.leetcode.top.interview.easy.LongestCommonPrefix 分治法处理，小任务处理组内最长长度
 *
 * @author wezhyn
 * @since 08.23.2020
 *
 */
class LongestCommonPrefix {
    fun longestCommonPrefix(strs: Array<String>): String {
        return backtracking(strs, 0)
    }

    private fun backtracking(strs: Array<String>, index: Int): String {
        if (strs.isEmpty()) {
            return ""
        }
        var cur: Char? = null
        for (s in strs) {
            if (index >= s.length) {
                return ""
            }
            if (cur == null) {
                cur = s[index]
            } else {
                if (cur != s[index]) {
                    return ""
                }
            }
        }
        return cur!! + backtracking(strs, index + 1)
    }
}